crates/build/src/docker.rs

```diff
@@ -0,0 +1,535 @@
+
+    let container_option = containers.first();
+
+    let mount = Mount {
+        target: Some(String::from("/contract")),
+        source: Some(
+            host_folder
+                .to_str()
+                .context("Cannot convert path to string.")?
+                .to_string(),
+        ),
+        typ: Some(MountTypeEnum::BIND),
+        ..Default::default()
+    };
+    let host_cfg = Some(HostConfig {
+        mounts: Some(vec![mount]),
+        ..Default::default()
+    });
+
+    let mut labels = HashMap::new();
+    labels.insert("cmd_digest".to_string(), digest.to_string());
+    let config = Config {
+        image: Some(build_image.id.clone()),
+        entrypoint: Some(entrypoint),
+        cmd: None,
+        labels: Some(labels),
+        host_config: host_cfg,
+        attach_stderr: Some(true),
+        tty: Some(true),
+        ..Default::default()
+    };
+    let options = Some(CreateContainerOptions {
+        name: container_name.as_str(),
+        platform: Some("linux/amd64"),
+    });
+
+    let container_id = match container_option {
+        Some(container) => {
+            container
+                .id
+                .clone()
+                .context("Container does not have an ID")?
+        }
+        None => client.create_container(options, config).await?.id,
+    };
+
+    client
+        .start_container::<String>(&container_id, None)
+        .await?;
+
+    maybe_println!(
+        verbosity,
+        " {} {}\n {}",
+        format!("{build_steps}").bold(),
+        "Started the build inside the container"
+            .bright_green()
+            .bold(),
+        "You can close this terminal session. The execution will be finished in the background"
+    );
+
+    let options = Some(WaitContainerOptions {
+        condition: "not-running",
+    });
+
+    let mut wait_stream = client.wait_container(&container_id, options);
+    let handle_error = |r: Result<ContainerWaitResponse, Error>| -> Result<()> {
+        let response = match r {
+            Ok(v) => v,
+            Err(e) => {
+                // sometimes we cannot remove the file due to privileged access
+                let _ = std::fs::remove_file(build_result_path);
+                anyhow::bail!(
+                    "{}. Execution failed! Check logs for more details",
+                    e.to_string()
+                )
+            }
+        };
+        if response.status_code != 0 {
+            anyhow::bail!("Execution failed! Status code: {}. Check the container's logs for more details", response.status_code);
+        }
+        Ok(())
+    };
+    if verbosity.is_verbose() {
+        let spinner_style =
+            ProgressStyle::with_template(" {spinner:.cyan.bold} {wide_msg}")?
+                .tick_chars("⠁⠂⠄⡀⢀⠠⠐⠈ ");
+
+        let pb = ProgressBar::new(1000);
+        pb.enable_steady_tick(Duration::from_millis(100));
+        pb.set_style(spinner_style);
+        pb.set_message("Build is being executed...");
```

Comment:
I think it would be much better to capture the output of the build via the logs and pipe it to the console, in order that the build looks similar to a "local" build. Also will make it easier to troubleshoot what is going on inside the container.

Reply:
There is no output as the docker build is carried out with `--output-json` flag

Reply:
An idea would be to change all the progress output to `stderr` (as cargo does) and then we can still show output even for `--output-json`. Anyway this is a nice to have.